File: pg.py
from psycopg2 import connect
from time import time as clock


# region: foundation

def timeit(name, f, n=5000, k=3):
    print(name)
    for j in xrange(k):
        t = clock()
        for i in xrange(n):
            f()
        t = clock() - t
        print(' #%d => %.1frps' % (j + 1, n / t))


# region: config

session = connect('dbname=main')


# region: models

class UserItem(object):

    def __init__(self, user_id, display_name, mail, phone, roles, status):
        self.user_id = user_id
        self.display_name = display_name
        self.mail = mail
        self.phone = phone
        self.roles = roles
        self.status = status


class User(object):

    def __init__(self, user_id='', status='1', customer_id='',
                 display_name='', mail='', phone1='', phone2='',
                 roles=''):
        self.user_id = user_id
        self.status = status
        self.customer_id = customer_id or ''
        self.display_name = display_name
        self.mail = mail
        self.phone1 = phone1
        self.phone2 = phone2
        self.roles = roles


# region: repository

def search_user(customer_id=None):
    c = session.cursor()
    c.callproc('membership.search_user', (customer_id,))
    return tuple(UserItem(*i) for i in c.fetchall())

def count_users(customer_id):
    c = session.cursor()
    c.callproc('membership.count_users', (customer_id,))
    return c.fetchone()[0]

def get_user(user_id):
    c = session.cursor()
    c.callproc('membership.get_user', (user_id,))
    r = c.fetchone()
    if not r:
        return None
    return User(*r)


# region: test cases

timeit('get user', lambda: get_user('1'))
timeit('search user', lambda: search_user('10'))
timeit('count users', lambda: count_users('10'))

# region: close everything
session.close()

File: pg.lua
local clock = require('socket').gettime
local driver = require 'luasql.postgres'
local assert, format = assert, string.format


-- region: foundation

local Session = {__index = {
    fetchscalar = function(self, stmt, ...)
        local c = assert(self.c:execute(format(stmt, ...)))
        return c:fetch()
    end,
    fetchone = function(self, stmt, ...)
        local c = assert(self.c:execute(format(stmt, ...)))
        return c:fetch({}, 'a')
    end,
    fetchall = function(self, stmt, ...)
        local c = assert(self.c:execute(format(stmt, ...)))
        local i = 1
        local t = {}
        while true do
            local r = c:fetch({}, 'a')
            if not r then
                break
            end
            t[i] = r
            i = i + 1
        end
        return t
    end,
    escape = function(self, s)
        return self.c:escape(s)
    end,
    close = function(self)
        return self.c:close()
    end
}}

local function timeit(name, f, n, k)
    print(name)
    n = n or 5000
    k = k or 3
    for j = 1, k do
        local t = clock()
        for i = 1, n do
            f()
        end
        t = clock() - t
        print(format(' #%d => %.1frps', j, n / t))
    end
end

-- region: config

local env = assert(driver.postgres())
local session = setmetatable({c=assert(env:connect('dbname=main'))}, Session)
session.c:execute('SET default_transaction_read_only = on;')

-- region: repository

local function search_user(customer_id)
    return session:fetchall(
        "SELECT membership.search_user('%s')",
        customer_id)
end

local function get_user(user_id)
    return session:fetchone(
        "SELECT membership.get_user('%s')",
        user_id)
end

local function count_users(customer_id)
    return session:fetchscalar(
        "SELECT membership.count_users('%s')",
        customer_id)
end

-- region: test cases

timeit('get user', function()
    return get_user('1')
end)

timeit('search user', function()
    return search_user('10')
end)

timeit('count users', function()
    return count_users('10')
end)

-- region: close everything
session:close()
env:close()
